From e6298b787b3510c295dd0c9276194b3578f3cf09 Mon Sep 17 00:00:00 2001
From: Rakshith Suresh Patkar <rpatkar@codeaurora.org>
Date: Fri, 11 May 2018 18:28:59 +0530
Subject: qcacld-3.0: Fix OOB access in htt_rx_ring_fill_n

Validate num and idx variables to avoid OOB access.

Change-Id: I920a3cd12744055cfc8315e3b16f8564a3cf9683
CRs-Fixed: 2225604
---
 core/dp/htt/htt_rx.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/core/dp/htt/htt_rx.c b/core/dp/htt/htt_rx.c
index 8a04fae..2315d53 100644
--- a/core/dp/htt/htt_rx.c
+++ b/core/dp/htt/htt_rx.c
@@ -475,6 +475,15 @@ static int htt_rx_ring_fill_n(struct htt_pdev_t *pdev, int num)
 	int num_alloc = 0;
 
 	idx = *(pdev->rx_ring.alloc_idx.vaddr);
+
+	if ((idx < 0) || (idx > pdev->rx_ring.size_mask) ||
+	    (num > pdev->rx_ring.size))  {
+		QDF_TRACE(QDF_MODULE_ID_HTT,
+			  QDF_TRACE_LEVEL_ERROR,
+			  "%s:rx refill failed!", __func__);
+		return filled;
+	}
+
 	if (qdf_mem_smmu_s1_enabled(pdev->osdev) && pdev->is_ipa_uc_enabled) {
 		mem_map_table = qdf_mem_map_table_alloc(num);
 		if (!mem_map_table) {
-- 
cgit v1.1

